### ISSUE TITLE
Multiple projects management

### ISSUE THREAD

**Issue by kpavel** [apache/incubator-openwhisk-wskdeploy#730]
Currently it is impossible to use wskdeploy to  deploy OW assets (action, package...etc) of the same signature from different manifests having different project names. The problem is that wskdeploy overrides the "managed project" annotation. The workaround could be to use different package names, which results in assets duplication. Duplication is undesirable, because it complicates maintenance. 
Introducing a feature allowing to manage OW assets by multiple projects would help to avoid it.
Suggestion:
- Use different annotation key instead of "manage" for the sake of backward compatibility. E.g. `managedList` holding an array of managed project names.
- On manifest deployment, if a deployed asset already exists in OW, check whether the manifest asset is identical to the existing one.
If not: override the existing asset with the manifest one (similar to the current functionality)
If yes: add the project specified in the manifest to the `managedList`.

**Comment by mrutkows** [comment]
@kpavel as discussed elsewhere... the plan would be to use URI "tags" where there is no key (each project name would be clearly part of the URI) along with domain/path to indicate that this is a "managed project" tag.

**Comment by lionelvillard** [comment]
Are your projects statically dependent to each other or not? If yes, then `dependencies` might solve your problem by putting the common assets into a "library manifest". Otherwise can you deploy each project to their own namespace?

**Comment by pritidesai** [comment]
so for now, instead of creating annotation as a JSON value, we can have an array of JSON blob, something similar to:

```
    "annotations": [
        {
            "key": "managed",
            "value": [
                {
                    "k1": "v1",
                    "k2": "v2"
                },
                {
                    "k1": "v1",
                    "k2": "v2"
                }
            ]
        },
```

thoughts?

**Comment by davidbreitgand** [comment]
@pritidesai This is exactly what we had in mind. @lionelvillard The functionality we want to achieve is that there will be only one copy of an annotated asset that when updated retains connections to the projects that it is managed by. It's our understanding that creating a library asset is a possible solution, but with the current implementation, if the user updates the library asset, she will need to call wskdeply for each project that has dependency on that asset and the asset will be overwritten N times in the DB, where N is the number of projects. Having an array of annotations, as @pritidesai suggests seems to be a faster and cleaner way of achieving the result. Thoughts?

**Comment by mrutkows** [comment]
@pritidesai if we move to a nested map, as we discussed yesterday within the annotations, I would prefer that we attempt to simulate a key-only "tag" which is URI based.  Where the URI includes information regarding the project and project hash.

**Comment by lionelvillard** [comment]
The idea of the explicit dependencies approach is to have only one project which include all your dependencies. Only one project needs to be deployed (and wskdeploy can compute an optimal deployment plan). 

Going back to the multiple projects management issue (in which dependencies are not explicitly specified and instead managed outside of wskdeploy), here I currently think what could go wrong and things to be aware of:

-  if asset a in Project A and asset b in Project B have the same name but not identical (different code, parameters, etc...) then in general b cannot be deployed, otherwise it would break A. 
-  consequently: if the purpose of B is to extends A (e.g. adding new actions to a package P), then both package descriptions P in A and B must be identical (same parameters, annotations, ...). 
- comparing feeds for equality required adding additional annotations to store feed parameters.
- comparing APIs should probably be done by comparing swagger specifications

Bottom line: 
- By default, deployment should fail when assets of the same name are different. A flag could be provided to override this behavior
- Asset equality must be formally defined to support the default behavior above.

**Comment by davidbreitgand** [comment]
@lionelvillard  I think that --override flag makes perfect sense. Also agreed that assets equality definition is required. Concerning explicit dependencies approach, assume that you have project A -> B and project C -> B, where -> means "depends on". For the sake of discussion, wskdeploy project == OW package. Furthermore, there is a an asset b in B. Now, suppose the user does wsk action update b blah-blah. How would that change propagate to  A and C? The dependencies are reflected only in the manifests of the respective projects, but nowhere in the OW data model, right? So, today the user will have to wskdeploy A and wskdeploy C. With multiple projects annotations on the asset, the user just does  wsk action update b carefree, because the dependencies of other projects on asset b are reflected in the annotations of the asset b.

**Comment by davidbreitgand** [comment]
@mrutkows Can you define the tag for @kpavel  and my education (you probably discussed that within the team, so sorry for asking again). Just want to make sure we're on the same page.

**Comment by lionelvillard** [comment]
@davidbreitgand in this model, there is only one project:
```
P --> A --> B
  |-> C -->  B
```

then it's up to wskdeploy to factor out B. When both A and C instantiates B with the same parameters, then all is fine. Otherwise wskdeploy can reject or decide to deploy B is two different packages (when there is no dynamic bindings) or two different namespaces (wskdeploy automatically handles authentication). 

As for `wsk action update b`, I suppose you also mean the user will then use `export`. In that case the expected behavior is to update the action in the manifest B. However as you said, the OW data model does not support dependencies so in order to propagate changes to A and C (like for instance if a parameter on b is defined in A) additional annotations must be added to encode such dependencies.

To be clear: I'm not against adding the `managedList` annotation, I'm just trying to contrast this approach with the `dependencies` approach. It seems to me that in order to let wskdeploy managing dependencies we will need more than the `managedList` annotation.

**Comment by mrutkows** [comment]
@davidbreitgand Using your example of, A -> B and project C -> B, my initial reaction is that I do not  support making any changes to project B's annotations.  As described, it seems that project B has value (as providing a library/shared asset) and should be marked public/shared by its owner (different owner than project A and project C).   From a security point of view (and existing dependency paradigms) we should never alter another person's assets even if we are only affecting their annotations.
